### PR TITLE
feat: New user panel: First phase

### PR DESCRIPTION
**Pull request by rique223**
## Jira epic: [WM-51](https://rocketchat.atlassian.net/browse/WM-51?atlOrigin=eyJpIjoiMjRhOGEyY2Y2MWEzNDQyNzk3Mjk3NDYxZmUwZmEzMmEiLCJwIjoiaiJ9)

Demo video: https://www.loom.com/share/210fcb90a4724ae39d59c8405b645814?sid=c66bd484-935a-403e-aa48-c0bfa7ecbfae

This PR describes the first phase of a complete refactor of the users page by:

- Dividing the page in 4 tabs:
    * All: Lists all users and changes the status column to `Registration status`;
    * Pending: Lists users that either havent been activated by an admin yet or that have been activated but did not login for the first time, has a `Pending action` column that lists which action the given user is pending and added two new buttons, `Activate` so that the admin can quickly activate an user without having to access inner menus and the `Resend welcome email` so that the user can resend the welcome email and remind users that they can login for the first time;
    * Active: Lists the active users (Users that have been activated and logged in for the first time);
    * Deactivated: Lists the deactivatedf users (Users that have logged in for the first time but for some reason where deactivated);

- Implementing a multi select roles filter;

- Refactoring the contextual bar by:
    * Changing the order of the inputs to a more logical one;
    * Adding the reason for joining info when the `Manually activate new users` setting is enabled;

- Refactoring the New user contextual bar by:
    * Changing the order of the inputs to a more logical one;
    * Changing the `verified` toggle label to `Mark email as verified` and adding an info icon that explains the functionality of that toggle;
    * Dividing the `Set random password and send by email` toggle into two radios, one for setting the password randomly and the other to set the password manually;
    * Hidding the password field behind the new set random password radios and only showing it when the `Set manually` radio has been checked;
    * Creating a `Confirm password` input;
    * Implementing the `PasswordVerifier` component in those password fields;
    * Implementing a `Hide additional fields` at the end of the contextual bar;
    * Changing the `Reset` and `Save` buttons to a single `Add user` button;
    * Changing the `invalid email address` error message to `Please ensure that the email address is valid`;
    * Implementing a post-save confirmation page that allows the admin to create more than one user in sequence;
    * Implementing verifications to make sure email related form inputs are disabled when the SMTP is not configured;
    * Implementing a verification to disable the `Mark email as verified` input and to warn the user that email verification is not needed in case the `Require email verification to login` setting is disabled;
    * Changing the `Only allow verified users to login / Make sure you have correct SMTP` to `Require email verification to login / Ensure SMTP is configured to enable`

- Refactoring the back-end by:
    * Creating a whole new endpoint called `users.list/:status` in order not to break probable unknown places that use the old endpoint;
    * Adding a `status` url param to the endpoint that receives the current tab and filters the returned users list to have only users that belong to that given tab;
    * Adding a `roles` query param to the endpoint that receives a list of role name strings and filters the returned users list to have only users that have at least one of those roles;
    * Adding the `lastLogin`, `type` and `reason` fields to the user objects;
    * Correctly implement the `view-full-other-user-info` setting on the users list;
    * Adding JSON Schema verification to the request params and implementing proper typing to those;
    * Refactoring the users.list endpoint typing to a more concise version of it´ 

## List of jira tasks closed in this PR and their respective PRs:

 [WM-117](https://rocketchat.atlassian.net/browse/WM-117?atlOrigin=eyJpIjoiMTQ3MGZlZjA4MmZiNGZhODljYWUzYzFmOTgyNWI1NzkiLCJwIjoiaiJ9) -> [feat: User Panel > Tab ALL](https://github.com/RocketChat/Rocket.Chat/pull/29622)

[WM-164](https://rocketchat.atlassian.net/browse/WM-164?atlOrigin=eyJpIjoiNDI5MzQ2ZjJjY2I5NGM0Yjk2ZDA3ZDZiNzIyYjAyYWEiLCJwIjoiaiJ9) -> [feat: pending tab new](https://github.com/RocketChat/Rocket.Chat/pull/30666)

[WM-86](https://rocketchat.atlassian.net/browse/WM-86?atlOrigin=eyJpIjoiYWQwYThlOTA2YjNiNDYyYmI5NTQwNDA0Mzc4ZTI2ZWUiLCJwIjoiaiJ9) -> [feat: Implement users page active tab](https://github.com/RocketChat/Rocket.Chat/pull/30242)

[WM-87](https://rocketchat.atlassian.net/browse/WM-87?atlOrigin=eyJpIjoiYWQ1MzYwZWQzNzUxNDUwMmJiNDBkODA3MDFmMTgyNzMiLCJwIjoiaiJ9) -> [feat: Implement users page deactivated tab](https://github.com/RocketChat/Rocket.Chat/pull/30532)

[WM-88](https://rocketchat.atlassian.net/browse/WM-88?atlOrigin=eyJpIjoiYjI3NTFkNDE3ODE5NDlkNGFjZTFkMjI1YjVlYWQ0NGEiLCJwIjoiaiJ9) -> [feat: New create user contextual bar](https://github.com/RocketChat/Rocket.Chat/pull/30582)
    
[WM-185](https://rocketchat.atlassian.net/browse/WM-185) -> [feat: endpoints to new users panel](https://github.com/RocketChat/Rocket.Chat/pull/30869)

Obs.: The contextual bar refactors and the filters are considered under the tab tasks.

## Further comments

[WM-51]: https://rocketchat.atlassian.net/browse/WM-51?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/meteor/client/components/UserCard/UserCardRoles.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ import UserCardInfo from './UserCardInfo';
 
 const UserCardRoles = ({ children }: { children: ReactNode }): ReactElement => (
 	<Box m='neg-x2'>
-		<UserCardInfo flexWrap='wrap' display='flex' flexShrink={0}>
+		<UserCardInfo mb={8} flexWrap='wrap' display='flex' flexShrink={0}>
```
Comment (ggazzo) [inline apps/meteor/client/components/UserCard/UserCardRoles.tsx:9]: set the margin inside the component

File: apps/meteor/client/views/admin/users/UsersTable/UsersTable.tsx [hunk 2]
```diff
@@ -17,86 +15,71 @@ import {
 } from '../../../../components/GenericTable';
 import { usePagination } from '../../../../components/GenericTable/hooks/usePagination';
 import { useSort } from '../../../../components/GenericTable/hooks/useSort';
+import useFilteredUsers from '../hooks/useFilteredUsers';
 import UsersTableRow from './UsersTableRow';
 
+// TODO: remove bots from pending users + add "resend email" function
```
Comment (hugocostadev) [inline apps/meteor/client/views/admin/users/UsersTable/UsersTable.tsx:21]: review and remove unnecessary TODOs

File: apps/meteor/client/views/admin/users/UsersTable/UsersTableRow.tsx [hunk 3]
```diff
@@ -1,35 +1,117 @@
...
+			delete: { label: { label: deleteUserAction.label, icon: deleteUserAction.icon }, action: deleteUserAction.action },
+		}),
+	};
+
+	// TODO: create action for this?
+	// TODO: implement logic
+	const handleResendWelcomeEmail = () => {
+		console.log('Welcome email resent');
+		dispatchToastMessage({ type: 'success', message: t('Welcome_email_resent') });
+	};
```
Comment (hugocostadev) [inline apps/meteor/client/views/admin/users/UsersTable/UsersTableRow.tsx:94]: This is not doing anything

Comment (hugocostadev) [reply apps/meteor/client/views/admin/users/UsersTable/UsersTableRow.tsx:94]: Remember to check SMTP settings to trigger this function

**Review by hugocostadev — CHANGES_REQUESTED** [review]

File: apps/meteor/app/api/server/v1/users.ts [hunk 4]
```diff
@@ -555,6 +559,125 @@ API.v1.addRoute(
...
+				case 'deactivated':
+					match = {
+						active: false,
+						lastLogin: { $exists: true },
+					};
+					break;
+				case 'pending':
+					match = {
+						lastLogin: { $exists: false },
+						type: { $ne: 'bot' },
```
[the remark this answers is not in the document]
Comment (hugocostadev) [reply apps/meteor/app/api/server/v1/users.ts]: Fixed

File: apps/meteor/app/api/server/v1/misc.ts [hunk 5]
```diff
@@ -634,9 +635,8 @@ API.v1.addRoute(
 	{ authRequired: true },
 	{
 		async get() {
-			const isMailURLSet = !(process.env.MAIL_URL === 'undefined' || process.env.MAIL_URL === undefined);
-			const isSMTPConfigured = Boolean(settings.get('SMTP_Host')) || isMailURLSet;
-			return API.v1.success({ isSMTPConfigured });
+			const smtpConfigured = isSMTPConfigured();
+			return API.v1.success({ smtpConfigured });
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/misc.ts:639]: This is a smol breaking change. We should keep same return value

File: apps/meteor/app/api/server/v1/users.ts [hunk 6]
```diff
@@ -555,6 +559,136 @@ API.v1.addRoute(
...
+					...match,
+				},
+				{
+					sort: actualSort,
+					skip: offset,
+					limit: count,
+					projection,
+				},
+			);
+
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/users.ts:658]: Use promise.all

File: apps/meteor/app/api/server/v1/users.ts [hunk 7]
```diff
@@ -555,6 +559,136 @@ API.v1.addRoute(
...
+	'users.sendWelcomeEmail',
+	{
+		authRequired: true,
+		validateParams: isUsersSendWelcomeEmailProps,
+	},
+	{
+		async post() {
+			const { email } = this.bodyParams;
+
+			try {
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/users.ts:682]: No need for the try catch since you're sending the same error received from the func.

File: apps/meteor/app/api/server/v1/misc.ts (hunk 5 above)
Comment (hugocostadev) [reply apps/meteor/app/api/server/v1/misc.ts]: Thanks for noticing that man!

File: apps/meteor/app/api/server/v1/misc.ts [hunk 8]
```diff
@@ -634,9 +635,8 @@ API.v1.addRoute(
 	{ authRequired: true },
 	{
 		async get() {
-			const isMailURLSet = !(process.env.MAIL_URL === 'undefined' || process.env.MAIL_URL === undefined);
-			const isSMTPConfigured = Boolean(settings.get('SMTP_Host')) || isMailURLSet;
-			return API.v1.success({ isSMTPConfigured });
+			const smtpConfigured = isSMTPConfigured();
+			return API.v1.success({ isSMTPConfigured: smtpConfigured });
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/misc.ts:639]:
```suggestion
			return API.v1.success({ isSMTPConfigured: isSMTPConfigured() });
```

**Review by hugocostadev — APPROVED** [review]

File: apps/meteor/app/api/server/v1/users.ts [hunk 9]
```diff
@@ -555,6 +559,131 @@ API.v1.addRoute(
 	},
 );
 
+API.v1.addRoute(
+	'users.list/:status',
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/users.ts:563]: Hey @MarcosSpessatto , was thinking on renaming this one to `users.listByStatus` to follow the other conventions on the users file, and passing the status as a query param. 

Wdyt?

File: apps/meteor/app/api/server/v1/users.ts [hunk 10]
```diff
@@ -555,6 +559,131 @@ API.v1.addRoute(
...
+			};
+
+			if (roles?.length && !roles.includes('all')) {
+				match = {
+					...match,
+					roles: { $in: roles },
+				};
+			}
+
+			const result = await Users.findPaginated(
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/users.ts:647]: This is a nitpick so safe to ignore, but why not:
```suggestion
			const { cursor, totalCount } = await Users.findPaginated(
```

This would remove the `result.cursor.toArray` from L659, and would be similar to how we call findPaginated on most places

File: apps/meteor/app/api/server/v1/users.ts [hunk 11]
```diff
@@ -555,6 +559,131 @@ API.v1.addRoute(
...
+	{
+		authRequired: true,
+		validateParams: isUsersSendWelcomeEmailProps,
+	},
+	{
+		async post() {
+			const { email } = this.bodyParams;
+			await sendWelcomeEmail(email);
+
+			return API.v1.success({ success: true });
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/users.ts:682]: You can pass an empty object and the `.success` call will add success: true to it

File: apps/meteor/server/methods/sendWelcomeEmail.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+import * as Mailer from '../../app/mailer/server/api';
+import { settings } from '../../app/settings/server';
+import { isSMTPConfigured } from '../../app/utils/server/functions/isSMTPConfigured';
+
+export async function sendWelcomeEmail(to: string): Promise<void> {
+	if (typeof to !== 'string') {
+		throw new Meteor.Error('error-invalid-arguments', 'Invalid arguments', {
+			method: 'sendWelcomeEmail',
+		});
+	}
```
Comment (KevLehman) [inline apps/meteor/server/methods/sendWelcomeEmail.ts:13]: Not that I'm against validating, but if the endpoint is already validating this param is an string, is this extra check needed?

File: apps/meteor/server/methods/sendWelcomeEmail.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+		throw new Meteor.Error('error-invalid-user', 'Invalid user', {
+			method: 'sendWelcomeEmail',
+		});
+	}
+
+	try {
+		let html = '';
+		Mailer.getTemplate('Accounts_UserAddedEmail_Email', (template) => {
+			html = template;
+		});
```
Comment (KevLehman) [inline apps/meteor/server/methods/sendWelcomeEmail.ts:35]: Curious: will the callback on getTemplate be called always before Mailer.send needs the data?

File: apps/meteor/tests/end-to-end/api/01-users.js [hunk 14]
```diff
@@ -4229,4 +4229,179 @@ describe('[Users]', function () {
 			});
 		});
 	});
+
```
Comment (KevLehman) [inline apps/meteor/tests/end-to-end/api/01-users.js:4232]: Tests for the sendWelcomeEmail endp would be good too

File: apps/meteor/server/methods/sendWelcomeEmail.ts (hunk 12 above)
Comment (hugocostadev) [reply apps/meteor/server/methods/sendWelcomeEmail.ts:13]: Because this Fn can be used elsewhere

File: apps/meteor/server/methods/sendWelcomeEmail.ts (hunk 13 above)
Comment (hugocostadev) [reply apps/meteor/server/methods/sendWelcomeEmail.ts:35]: Yes, it will, settings are already "ready" let's say, also, the Mailer.getTemplate does not return a Promise, so we can't (it's going to have any effect) use await here

File: apps/meteor/app/api/server/v1/users.ts (hunk 9 above)
Comment (rique223) [reply apps/meteor/app/api/server/v1/users.ts:563]: @tapiarafael Could we get some input on the necessity of the status as a url param here please?

File: apps/meteor/tests/end-to-end/api/01-users.js (hunk 14 above)
Comment (hugocostadev) [reply apps/meteor/tests/end-to-end/api/01-users.js:4232]: Done!

File: apps/meteor/app/api/server/v1/users.ts (hunk 11 above)
Comment (hugocostadev) [reply apps/meteor/app/api/server/v1/users.ts]: Done! Thanks 🚀

File: apps/meteor/app/api/server/v1/users.ts (hunk 10 above)
Comment (hugocostadev) [reply apps/meteor/app/api/server/v1/users.ts]: Much better! Thanks man 👍

File: apps/meteor/app/api/server/v1/users.ts (hunk 9 above)
Comment (tapiarafael) [reply apps/meteor/app/api/server/v1/users.ts:563]: I don't see any problem in changing the endpoint to follow the convention. The front-end will just need to send the status via query param instead of sending it as URL param. @rique223

**Review by KevLehman — DISMISSED** [review]

**Review by hugocostadev — APPROVED** [review]

File: .changeset/serious-poets-smash.md [hunk 15]
```diff
@@ -0,0 +1,6 @@
+---
+"@rocket.chat/meteor": minor
+"@rocket.chat/rest-typings": minor
+---
```
Comment (tassoevan) [inline .changeset/serious-poets-smash.md:4]:
```suggestion
---
"@rocket.chat/meteor": minor
"@rocket.chat/rest-typings": minor
"@rocket.chat/core-typings": patch
"@rocket.chat/ui-client": patch
---
```

File: apps/meteor/app/api/server/v1/users.ts [hunk 16]
```diff
@@ -555,6 +559,130 @@ API.v1.addRoute(
...
+
+			if (sort?.status) {
+				actualSort.active = sort.status;
+			}
+
+			if (sort?.name) {
+				actualSort.nameInsensitive = sort.name;
+			}
+
+			let match;
```
Comment (tassoevan) [inline apps/meteor/app/api/server/v1/users.ts:599]: image
A `let` declaration without type implicitly working as `let match: any`. We don't like implicit `any`s.

**Review by tassoevan — CHANGES_REQUESTED** [review]

File: apps/meteor/app/api/server/v1/users.ts (hunk 16 above)
Comment (hugocostadev) [reply apps/meteor/app/api/server/v1/users.ts]: Indeed! Thanks Tasso!  I've added the proper types 🚀

**Review by hugocostadev — APPROVED** [review]

File: apps/meteor/app/api/server/v1/users.ts (hunk 16 above)
Comment (KevLehman) [reply apps/meteor/app/api/server/v1/users.ts]: holy cow, this image covered my whole screen lol

File: packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+import type { PaginatedRequest } from '../../helpers/PaginatedRequest';
+
+const ajv = new Ajv({
+	coerceTypes: true,
+});
+
+export type UsersListStatusParamsGET = {
+	status: 'active' | 'all' | 'deactivated' | 'pending';
+	roles: string[];
+	searchTerm: string;
```
Comment (KevLehman) [inline packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts:12]: Your properties here are required, but on the AJV schema they're not. Which one is the right one?

Comment (hugocostadev) [reply packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts:12]: They are not required, let me fix

Comment (hugocostadev) [reply packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts:12]: Thanks 🚀

Comment (hugocostadev) [reply packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts:12]: Sorry, `status` is required

Comment (KevLehman) [reply packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts:12]: Oh, if its required, then add is as `required: ['status']` on the ajv schema, before additionalProperties

assuming that's the only required one, remember to update the roles & searchTerm props on this type to put them as optional

File: packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+} & PaginatedRequest;
```
Comment (KevLehman) [inline packages/rest-typings/src/v1/users/UsersListStatusParamsGET.ts:13]: PaginatedRequest is a generic type, you can put `PaginatedRequest<YourType>`. It works almost the same, but...

File: apps/meteor/client/views/admin/users/AdminUserCreated.tsx [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+import React, { useCallback } from 'react';
+import { useTranslation } from 'react-i18next';
+
+import { ContextualbarEmptyContent } from '../../../components/Contextualbar';
+
+const AdminUserCreated = ({ uid }: { uid: string }) => {
+	const { t } = useTranslation();
+	const router = useRouter();
+
+	const goToUser = useCallback((id) => router.navigate(`/admin/users/info/${id}`), [router]);
```
Comment (tassoevan) [inline apps/meteor/client/views/admin/users/AdminUserCreated.tsx:12]: There is no strong re-computation happening to justify memoizing this function.

File: apps/meteor/client/views/admin/users/AdminUserCreated.tsx [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+
+	return (
+		<>
+			<ContextualbarEmptyContent icon='user' title={t('You_have_created_user', { count: 1 })} />
+			<ContextualbarFooter>
+				<ButtonGroup stretch>
+					<Button type='reset' w='50%' onClick={() => router.navigate(`/admin/users/new`)}>
+						{t('Add_more_users')}
+					</Button>
+					<Button primary w='50%' onClick={() => goToUser(uid)}>
```
Comment (tassoevan) [inline apps/meteor/client/views/admin/users/AdminUserCreated.tsx:22]:
```suggestion
					<Button primary w='50%' onClick={() => router.navigate(`/admin/users/info/${uid}`)}>
```

**Review by tassoevan — CHANGES_REQUESTED** [review]

File: apps/meteor/client/views/admin/users/AdminUserForm.tsx [hunk 21]
```diff
@@ -127,25 +135,35 @@ const UserForm = ({ userData, onReload, ...props }: AdminUserFormProps) => {
...
+
+		if (!isNewUserPage && userData?._id) {
 			return handleUpdateUser.mutateAsync({ userId: userData?._id, data: userFormData });
 		}
-		return handleCreateUser.mutateAsync(userFormData);
+
+		return handleCreateUser.mutateAsync(
+			{ ...userFormData, fields: '' },
+			{
+				onSuccess: () =>
```
Comment (tassoevan) [inline apps/meteor/client/views/admin/users/AdminUserForm.tsx:161]: Is this overriding or adding to the `onSuccess` clause declare on the `useMutation` call? Regardless, why is this here instead of there?

**Review by tassoevan — CHANGES_REQUESTED** [review]

File: apps/meteor/client/views/admin/users/AdminUserForm.tsx (hunk 21 above)
Comment (rique223) [reply apps/meteor/client/views/admin/users/AdminUserForm.tsx:161]: Neither, the onSuccess of mutateAsync will only run when this specific mutate function runs while the onSucces of the useMutation will run after any successfull mutate. In hindsight the onSuccess of the useMutation is a better place to have this invalidation. Will fix in a sec.

**Review by tapiarafael — APPROVED** [review]
